File: bindings/php/src/Essential.php
<?php

declare(strict_types=1);

namespace STRling;

require_once __DIR__ . '/Core/Nodes/Node.php';
require_once __DIR__ . '/Simply.php';

use STRling\Core\Nodes;

/**
 * STRling Essential — compatibility lexical-shape patterns for common string
 * formats. They do not establish semantic validity or standards conformance.
 *
 * Each helper composes the existing Pattern AST so the compiled output
 * flows through the standard pipeline and no raw regex leaks into the
 * public API.
 */
final class Essential
{
    private static function letterItems(): array
    {
        return [
            new Nodes\Range(from: 'A', to: 'Z'),
            new Nodes\Range(from: 'a', to: 'z'),
        ];
    }

    private static function digitItems(): array
    {
        return [new Nodes\Escape('digit')];
    }

    private static function hexItems(): array
    {
        return [
            new Nodes\Range(from: 'A', to: 'F'),
            new Nodes\Range(from: 'a', to: 'f'),
            new Nodes\Range(from: '0', to: '9'),
        ];
    }

    private static function charsItems(string $s): array
    {
        $out = [];
        for ($i = 0; $i < strlen($s); $i++) {
            $out[] = new Nodes\Literal($s[$i]);
        }
        return $out;
    }

    private static function classOf(array $items, int $min, ?int $max): Pattern
    {
        $cc = new Nodes\CharacterClass(negated: false, members: $items);
        if ($min === 1 && $max === 1) {
            return new Pattern($cc);
        }
        $q = new Nodes\Quantifier(
            target: $cc,
            min: $min,
            max: $max,
            greedy: true,
            lazy: false,
            possessive: false
        );
        return new Pattern($q);
    }

    private static function digN(int $min, ?int $max): Pattern
    {
        return self::classOf(self::digitItems(), $min, $max);
    }

    private static function hexN(int $min, ?int $max): Pattern
    {
        return self::classOf(self::hexItems(), $min, $max);
    }

    private static function lettersN(int $min, ?int $max): Pattern
    {
        return self::classOf(self::letterItems(), $min, $max);
    }

    private static function lit(string $s): Pattern
    {
        return new Pattern(new Nodes\Literal($s));
    }

    private static function opt(Pattern $body): Pattern
    {
        $grouped = new Nodes\Group(capturing: false, body: $body->getNode());
        $q = new Nodes\Quantifier(
            target: $grouped,
            min: 0,
            max: 1,
            greedy: true,
            lazy: false,
            possessive: false
        );
        return new Pattern($q);
    }

    private static function altOf(array $branches): Pattern
    {
        $nodes = array_map(fn(Pattern $p) => $p->getNode(), $branches);
        return new Pattern(new Nodes\Alternation(alternatives: $nodes));
    }

    private static function seqOf(array $parts): Pattern
    {
        return Simply::merge(...$parts);
    }

    /**
     * Matches the legacy email-like lexical shape; RFC 5322 conformance is not claimed.
     */
    public static function email(): Pattern
    {
        $local  = self::classOf([...self::letterItems(), ...self::digitItems(), ...self::charsItems('._%+-')], 1, null);
        $domain = self::classOf([...self::letterItems(), ...self::digitItems(), ...self::charsItems('.-')], 1, null);
        $tld    = self::lettersN(2, null);
        return self::seqOf([$local, self::lit('@'), $domain, self::lit('.'), $tld]);
    }

    /**
     * Matches the legacy HTTP(S) URL-like lexical shape; RFC 3986 conformance is not claimed.
     */
    public static function url(): Pattern
    {
        $base     = [...self::letterItems(), ...self::digitItems(), ...self::charsItems("/_-.~%&=:@!$'()*+,;")];
        $withQ    = [...$base, ...self::charsItems('?')];
        $withFrag = [...$withQ, ...self::charsItems('#')];

        $scheme   = self::seqOf([self::lit('http'), self::opt(self::lit('s'))]);
        $host     = self::classOf([...self::letterItems(), ...self::digitItems(), ...self::charsItems('.-')], 1, null);
        $port     = self::opt(self::seqOf([self::lit(':'), self::digN(1, null)]));
        $path     = self::opt(self::seqOf([self::lit('/'), self::classOf($base, 0, null)]));
        $query    = self::opt(self::seqOf([self::lit('?'), self::classOf($withQ, 0, null)]));
        $fragment = self::opt(self::seqOf([self::lit('#'), self::classOf($withFrag, 0, null)]));
        return self::seqOf([$scheme, self::lit('://'), $host, $port, $path, $query, $fragment]);
    }

    /**
     * Matches the RFC 9562 UUID text shape; version=4 constrains version/variant nibbles.
     */
    public static function uuid(int $version = 0): Pattern
    {
        $dash = fn() => self::lit('-');
        if ($version === 4) {
            $variant = self::classOf(self::charsItems('89ABab'), 1, 1);
            return self::seqOf([
                self::hexN(8, 8),  $dash(),
                self::hexN(4, 4),  $dash(),
                self::lit('4'), self::hexN(3, 3), $dash(),
                $variant, self::hexN(3, 3), $dash(),
                self::hexN(12, 12),
            ]);
        }
        return self::seqOf([
            self::hexN(8, 8),  $dash(),
            self::hexN(4, 4),  $dash(),
            self::hexN(4, 4),  $dash(),
            self::hexN(4, 4),  $dash(),
            self::hexN(12, 12),
        ]);
    }

    /**
     * Matches an IPv4-like or full-form IPv6 lexical shape; address validity is not claimed.
     */
    public static function ip(int $version = 0): Pattern
    {
        $ipv4 = fn() => self::seqOf([
            self::digN(1, 3), self::lit('.'),
            self::digN(1, 3), self::lit('.'),
            self::digN(1, 3), self::lit('.'),
            self::digN(1, 3),
        ]);
        $ipv6 = fn() => self::seqOf([
            self::hexN(1, 4), self::lit(':'),
            self::hexN(1, 4), self::lit(':'),
            self::hexN(1, 4), self::lit(':'),
            self::hexN(1, 4), self::lit(':'),
            self::hexN(1, 4), self::lit(':'),
            self::hexN(1, 4), self::lit(':'),
            self::hexN(1, 4), self::lit(':'),
            self::hexN(1, 4),
        ]);
        return match ($version) {
            4 => $ipv4(),
            6 => $ipv6(),
            default => self::altOf([$ipv4(), $ipv6()]),
        };
    }

    /**
     * Matches a timestamp-like lexical shape; RFC 3339 / ISO 8601 validity is not claimed.
     */
    public static function dateTime(): Pattern
    {
        $sign   = self::classOf(self::charsItems('+-'), 1, 1);
        $frac   = self::seqOf([self::lit('.'), self::digN(1, null)]);
        $offset = self::seqOf([$sign, self::digN(2, 2), self::lit(':'), self::digN(2, 2)]);
        return self::seqOf([
            self::digN(4, 4), self::lit('-'), self::digN(2, 2), self::lit('-'), self::digN(2, 2),
            self::lit('T'),
            self::digN(2, 2), self::lit(':'), self::digN(2, 2), self::lit(':'), self::digN(2, 2),
            self::opt($frac),
            self::opt(self::altOf([self::lit('Z'), $offset])),
        ]);
    }
}

File: bindings/rust/src/essential.rs
//! Standard-library compatibility lexical-shape patterns for common string
//! formats. They do not establish semantic validity or standards conformance.
//!
//! Each helper composes existing AST node constructors so the compiled output
//! flows through the standard pipeline and no raw regex leaks into the public
//! API.

use crate::core::nodes::*;

fn letter_items() -> Vec<ClassItem> {
    vec![
        ClassItem::Range(ClassRange { from_ch: "A".into(), to_ch: "Z".into() }),
        ClassItem::Range(ClassRange { from_ch: "a".into(), to_ch: "z".into() }),
    ]
}

fn digit_items() -> Vec<ClassItem> {
    vec![ClassItem::Esc(ClassEscape { escape_type: "d".into(), property: None })]
}

fn hex_items() -> Vec<ClassItem> {
    vec![
        ClassItem::Range(ClassRange { from_ch: "A".into(), to_ch: "F".into() }),
        ClassItem::Range(ClassRange { from_ch: "a".into(), to_ch: "f".into() }),
        ClassItem::Range(ClassRange { from_ch: "0".into(), to_ch: "9".into() }),
    ]
}

fn chars_items(s: &str) -> Vec<ClassItem> {
    s.chars()
        .map(|c| ClassItem::Char(ClassLiteral { ch: c.to_string() }))
        .collect()
}

fn class_of(items: Vec<ClassItem>, min: i32, max: Option<i32>) -> Node {
    let class = Node::CharacterClass(CharacterClass { negated: false, items });
    if min == 1 && max == Some(1) {
        class
    } else {
        let max_bound = match max {
            Some(n) => MaxBound::Finite(n),
            None => MaxBound::Infinite("Inf".to_string()),
        };
        Node::Quantifier(Quantifier {
            target: QuantifierTarget { child: Box::new(class) },
            min,
            max: max_bound,
            mode: "Greedy".to_string(),
            greedy: true,
            lazy: false,
            possessive: false,
        })
    }
}

fn dig_n(min: i32, max: Option<i32>) -> Node { class_of(digit_items(), min, max) }
fn hex_n(min: i32, max: Option<i32>) -> Node { class_of(hex_items(), min, max) }
fn letters_n(min: i32, max: Option<i32>) -> Node { class_of(letter_items(), min, max) }

fn lit(s: &str) -> Node {
    Node::Literal(Literal { value: s.to_string() })
}

fn seq(parts: Vec<Node>) -> Node {
    Node::Sequence(Sequence { parts })
}

fn opt(node: Node) -> Node {
    let wrapped = Node::Group(Group {
        capturing: false,
        body: Box::new(node),
        name: None,
        atomic: None,
    });
    Node::Quantifier(Quantifier {
        target: QuantifierTarget { child: Box::new(wrapped) },
        min: 0,
        max: MaxBound::Finite(1),
        mode: "Greedy".to_string(),
        greedy: true,
        lazy: false,
        possessive: false,
    })
}

fn alt(branches: Vec<Node>) -> Node {
    Node::Alternation(Alternation { branches })
}

fn concat(a: Vec<ClassItem>, b: Vec<ClassItem>) -> Vec<ClassItem> {
    let mut out = a;
    out.extend(b);
    out
}

/// Matches the legacy email-like lexical shape; RFC 5322 conformance is not claimed.
pub fn email() -> Node {
    let local = class_of(
        concat(concat(letter_items(), digit_items()), chars_items("._%+-")),
        1, None);
    let domain = class_of(
        concat(concat(letter_items(), digit_items()), chars_items(".-")),
        1, None);
    let tld = letters_n(2, None);
    seq(vec![local, lit("@"), domain, lit("."), tld])
}

/// Matches the legacy HTTP(S) URL-like lexical shape; RFC 3986 conformance is not claimed.
pub fn url() -> Node {
    let base = concat(concat(letter_items(), digit_items()), chars_items("/_-.~%&=:@!$'()*+,;"));
    let with_q = concat(base.clone(), chars_items("?"));
    let with_frag = concat(with_q.clone(), chars_items("#"));

    let scheme = seq(vec![lit("http"), opt(lit("s"))]);
    let host = class_of(concat(concat(letter_items(), digit_items()), chars_items(".-")), 1, None);
    let port = opt(seq(vec![lit(":"), dig_n(1, None)]));
    let path = opt(seq(vec![lit("/"), class_of(base, 0, None)]));
    let query = opt(seq(vec![lit("?"), class_of(with_q, 0, None)]));
    let fragment = opt(seq(vec![lit("#"), class_of(with_frag, 0, None)]));
    seq(vec![scheme, lit("://"), host, port, path, query, fragment])
}

/// Matches the RFC 9562 UUID text shape. Pass `4` to constrain the version and
/// variant nibbles, or `0` to leave fields uninterpreted.
pub fn uuid(version: i32) -> Node {
    let dash = || lit("-");
    if version == 4 {
        let variant = class_of(chars_items("89ABab"), 1, Some(1));
        return seq(vec![
            hex_n(8, Some(8)), dash(),
            hex_n(4, Some(4)), dash(),
            lit("4"), hex_n(3, Some(3)), dash(),
            variant, hex_n(3, Some(3)), dash(),
            hex_n(12, Some(12)),
        ]);
    }
    seq(vec![
        hex_n(8, Some(8)), dash(),
        hex_n(4, Some(4)), dash(),
        hex_n(4, Some(4)), dash(),
        hex_n(4, Some(4)), dash(),
        hex_n(12, Some(12)),
    ])
}

/// Matches an IPv4-like or full-form IPv6 lexical shape; address validity is not claimed.
/// Pass `4` or `6` for family-specific matching, or `0` for either family.
pub fn ip(version: i32) -> Node {
    let ipv4 = || seq(vec![
        dig_n(1, Some(3)), lit("."),
        dig_n(1, Some(3)), lit("."),
        dig_n(1, Some(3)), lit("."),
        dig_n(1, Some(3)),
    ]);
    let ipv6 = || seq(vec![
        hex_n(1, Some(4)), lit(":"),
        hex_n(1, Some(4)), lit(":"),
        hex_n(1, Some(4)), lit(":"),
        hex_n(1, Some(4)), lit(":"),
        hex_n(1, Some(4)), lit(":"),
        hex_n(1, Some(4)), lit(":"),
        hex_n(1, Some(4)), lit(":"),
        hex_n(1, Some(4)),
    ]);
    match version {
        4 => ipv4(),
        6 => ipv6(),
        _ => alt(vec![ipv4(), ipv6()]),
    }
}

/// Matches a timestamp-like lexical shape; RFC 3339 / ISO 8601 validity is not claimed.
pub fn date_time() -> Node {
    let sign = class_of(chars_items("+-"), 1, Some(1));
    let frac = seq(vec![lit("."), dig_n(1, None)]);
    let offset = seq(vec![sign, dig_n(2, Some(2)), lit(":"), dig_n(2, Some(2))]);
    seq(vec![
        dig_n(4, Some(4)), lit("-"), dig_n(2, Some(2)), lit("-"), dig_n(2, Some(2)),
        lit("T"),
        dig_n(2, Some(2)), lit(":"), dig_n(2, Some(2)), lit(":"), dig_n(2, Some(2)),
        opt(frac),
        opt(alt(vec![lit("Z"), offset])),
    ])
}
